recipe/meta.yaml

```
@@ -49,18 +45,84 @@ requirements:
     - pytorch
     - pytorch =*=cuda*
     - setuptools
-  run:
-    - einops
-    - python
-    - pytorch =*=cuda*
 
-test:
-  imports:
-    - flash_attn
-  commands:
-    - pip check
-  requires:
-    - pip
+outputs:
+  - name: flash-attn
+    requirements:
+      build:
+        - {{ compiler('cxx') }} # needed for DSO checker
+        - {{ stdlib('c') }}     # needed for DSO checker
+      host:
+        - cuda-version {{ cuda_compiler_version }}  # same cuda for host and build
+        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
+        - python
+        - pytorch =*=cuda*
```

The CUDA `*-dev` libraries have `run_exports`, which ensure their runtime libraries are added automatically with the appropriate pinnings. So suggesting including those here so the `flash-attn` package behaves the same

```suggestion
      host:
        - cuda-version {{ cuda_compiler_version }}  # same cuda for host and build
        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
        - libcublas-dev    # [(cuda_compiler_version or "").startswith("12")]
        - libcurand-dev    # [(cuda_compiler_version or "").startswith("12")]
        - libcusolver-dev  # [(cuda_compiler_version or "").startswith("12")]
        - libcusparse-dev  # [(cuda_compiler_version or "").startswith("12")]
        - python
        - libtorch         # required until pytorch run_exports libtorch
        - pytorch
        - pytorch =*=cuda*
```

---

```
WARNING (flash-attn): dso library package conda-forge/linux-64::libcurand==10.3.1.50=hd3aeb46_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (flash-attn): dso library package conda-forge/linux-64::libcublas==12.0.1.189=hd3aeb46_3 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (flash-attn): dso library package conda-forge/linux-64::libcusparse==12.0.0.76=hd3aeb46_2 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (flash-attn): dso library package conda-forge/linux-64::libcusolver==11.4.2.57=hd3aeb46_2 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

These libraries don't seem to be needed. Are they being statically linked?

---

Could also be that we forgot to add the appropriate linker args since we are rolling our own setup.py

---

@weiji14, what's the reason that this recipe depends on cublas-dev, but doesn't link to it? The pytorch CUDAExtension helper only adds links to pytorch libraries automatically, if these extensions link to cublas, then we need to list it e.g. https://github.com/carterbox/torch-radon/blob/05d83e2473ec238b8ae537a6e9b803fc83d49965/setup.py#L46

---

https://setuptools.pypa.io/en/latest/userguide/ext_modules.html#extension-api-reference

---

I added `libcublas-dev` in https://github.com/conda-forge/staged-recipes/pull/26239/commits/4d8b37c5bfcbdeb2a991f7a85588689aebf96470, see also previous thread discussion at https://github.com/conda-forge/staged-recipes/pull/26239#discussion_r1591726256. I think we needed `cusparse-dev` for Aten, unsure about `cublas-dev`...

---

This probably comes from PyTorch headers. Unfortunately ever case I've seen where something uses PyTorch at build time it `#include`s headers from PyTorch, which include these various CUDA libraries. So they need to be satisfied for the `#include`s of PyTorch to work

---

https://github.com/Dao-AILab/flash-attention/blob/478ee666cccbd1b8f63648633003059a8dc6827d/csrc/fused_dense_lib/fused_dense_cuda.cu#L10

At least blas is a direct dependency of these extensions.

---

For my sanity, I need to test with `-Wl,--no-undefined` or something like that.